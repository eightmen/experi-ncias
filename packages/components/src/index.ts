export * from './Box'
export * from './Flex'
export * from './Grid'
export * from './Button'
export * from './Link'
export * from './Paragraph'
export * from './Text'
export * from './Heading'
export * from './Image'
export * from './Card'
export * from './Label'
export * from './Input'
export * f