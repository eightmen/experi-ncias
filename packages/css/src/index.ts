
import type {
  CSSObject,
  ThemeUIStyleObject,
  ThemeDerivedStyles,
  Theme,
  ThemeUICSSObject,
} from './types'

export * from './types'

/**
 * Allows for nested scales with shorthand values
 * @example
 * {
 *   colors: {
 *     primary: { __default: '#00f', light: '#33f' }
 *   }
 * }
 * css({ color: 'primary' }); // { color: '#00f' }
 * css({ color: 'primary.light' }) // { color: '#33f' }
 */
export const THEME_UI_DEFAULT_KEY = '__default'

const hasDefault = (
  x: unknown
): x is { [THEME_UI_DEFAULT_KEY]: string | number } => {
  return typeof x === 'object' && x !== null && THEME_UI_DEFAULT_KEY in x
}

/**
 * Extracts value under path from a deeply nested object.
 * Used for Themes, variants and Theme UI style objects.
 * Given a path to object with `__default` key, returns the value under that key.
 *
 * @param obj a theme, variant or style object
 * @param path path separated with dots (`.`)
 * @param fallback default value returned if get(obj, path) is not found
 */
export function get(
  obj: object,
  path: string | number | undefined,
  fallback?: unknown,
  p?: number,
  undef?: unknown
): any {
  const pathArray = path && typeof path === 'string' ? path.split('.') : [path]
  for (p = 0; p < pathArray.length; p++) {
    obj = obj ? (obj as any)[pathArray[p]!] : undef
  }

  if (obj === undef) return fallback

  return hasDefault(obj) ? obj[THEME_UI_DEFAULT_KEY] : obj
}

export const getObjectWithVariants = (obj: any, theme: Theme): CSSObject => {
  if (obj && obj['variant']) {
    let result: CSSObject = {}
    for (const key in obj) {
      const x = obj[key]
      if (key === 'variant') {
        const val = typeof x === 'function' ? x(theme) : x
        const variant = getObjectWithVariants(get(theme, val as string), theme)
        result = { ...result, ...variant }
      } else {
        result[key] = x as CSSObject
      }
    }