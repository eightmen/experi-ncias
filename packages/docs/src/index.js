// @ts-check

/** @jsx jsx */
import { jsx } from 'theme-ui'

import Layout from './components/layout'

export { default as Banner } from './components/banner'
export { default as Tiles } from './components/tiles'
export { default as Cards } from './components/cards'
export { default as Note } from './components/note'

/**
 * @param {import("gatsby").WrapPageElementBrowserArgs} props
 */
export const WrapPageElement = ({ element, props }) => (
  <Layout {...props} children={element} />
)

/**
 * @see https://docsearch.algolia.com/
 * @param {import("gatsby").RenderBodyArgs} args
 */
export const setDocSearchComponents = ({
  setHeadComponents,
  setPostBodyComponents,
}) => {
  setHeadComponents([
    <link
      key="algolia-css"
      rel="stylesheet"
      href="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.css"
    />,
  ])
  setPostBodyComponents([
    <script
      key="algolia-cdn"
      type="text/javascript"
      src="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.js"
    ></script>,
    <script
      key="algolia-script"
      type="text/javascript"
      dangerouslySetInnerHTML={{
        __html: `
          var lastSearchInput = null; 
          var observer = new MutationObserver(function () {
            var searchSelector = "#algolia-docs-search";
            var searchInput = document.querySelector(searchSelector);

            if (searchInput && (lastSearchInput !== searchInput)) {
              docsearch({
                apiKey: "84ed820927eee5fa5018c9f1abe70390",
                indexName: "theme-ui",
                inputSelector: searchSelector,
                debug: true,
                transformData: function(hits) {
                  for (const hit of hits) {
                    hit.url = hit.url.replace(
                      "https://theme-ui.com",
                      window.location.origin
                    )
                  }
                }
              })
              lastSearchInput = searchInput;
            }
          });

          document.addEventListener("DOMContentLoaded", function() {
            observer.observe(document, {
              childList: true,
              subtree: true
            });
          });        
      `,
      }}
    />,
  ])
}
