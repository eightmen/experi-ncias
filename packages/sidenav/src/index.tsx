
/** @jsx jsx */
import { jsx, ThemeProvider, Theme, ThemeStyles, Global } from 'theme-ui'
import { MDXProvider, MDXProviderComponents } from '@mdx-js/react'
import React, {
  useState,
  FunctionComponent,
  EventHandler,
  MouseEvent,
  forwardRef,
  ReactNode,
  ReactComponentElement,
  ReactElement,
} from 'react'
import merge from 'deepmerge'

const createNestedLinks = (
  children: ReactNode,
  depth: number = 0
): ReactNode[] =>
  React.Children.toArray(children).reduce<ReactNode[]>((acc, child) => {
    if (!React.isValidElement(child)) {
      return acc
    }
    const type: string = child.props && child.props.mdxType
    if (!child.props || !child.props.children) return acc
    if (type === 'a') return [...acc, child]
    if (depth > 0 && type === 'ul') {
      const last = acc[acc.length - 1]
      acc[acc.length - 1] = React.cloneElement(last as ReactElement, {
        links: createNestedLinks(child.props.children),
      })
      return acc
    }
    return [...acc, ...createNestedLinks(child.props.children, depth + 1)]
  }, [])

const flattenLinks = (children: ReactNode): ReactNode[] =>
  React.Children.toArray(children).reduce<ReactNode[]>((acc, child) => {
    if (React.isValidElement(child)) {
      if (child.props && child.props.mdxType === 'a') {
        return [...acc, child]
      }
      if (!child.props || !child.props.children) return acc
      return React.Children.toArray([
        ...acc,
        ...flattenLinks(child.props.children),
      ])
    }
    return acc
  }, [])

interface OverlayProps {
  onClick?: EventHandler<MouseEvent<HTMLDivElement>>
}
const Overlay = ({ onClick }: OverlayProps) => (
  <React.Fragment>
    <div
      onClick={onClick}
      sx={{
        position: 'fixed',
        top: 0,
        right: 0,
        bottom: 0,
        left: 0,
      }}
    />
    <Global
      styles={{
        body: {
          overflow: ['hidden', 'auto'],
        },
      }}
    />
  </React.Fragment>
)

export const Sidenav = forwardRef<
  HTMLDivElement,
  {
    open?: boolean
    components?: MDXProviderComponents
    styles?: ThemeStyles
    children?: ReactNode
  }
>(({ open, styles = {}, components = {}, ...props }, ref) => {
  return (
    <ThemeProvider
      theme={{
        styles: merge<Theme['styles']>(
          {
            ul: {
              listStyle: 'none',
              p: 0,
              m: 0,
              ul: {
                a: {
                  pl: 4,
                },
              },
            },
            a: {
              display: 'block',
              px: 2,
              py: 2,
              color: 'inherit',
              textDecoration: 'none',
              fontSize: 1,
              fontWeight: 'bold',
            },
          },
          styles
        ),
      }}
    >
      {open && <Overlay />}
      <MDXProvider components={components}>
        <div
          {...props}
          ref={ref}
          sx={{
            position: ['fixed', 'sticky'],
            top: 0,
            left: 0,
            bottom: [0, 'auto'],
            zIndex: 1,
            minWidth: 0,
            width: 256,
            maxHeight: ['100vh', 'none'],
            overflowX: 'visible',
            overflowY: 'auto',
            transition: 'transform .2s ease-out',
            transform: [open ? 'translateX(0)' : 'translate(-100%)', 'none'],
            bg: ['background', 'transparent'],
            WebkitOverflowScrolling: 'touch',
          }}
        />
      </MDXProvider>
    </ThemeProvider>
  )
})

export const AccordionButton = (props: {
  open: boolean
  pathname?: string
  href: string
  className?: string
  onClick: EventHandler<MouseEvent<HTMLButtonElement>>
}) => {
  const transform = props.open ? 'rotate(-180 8 8)' : 'rotate(0 8 8)'
  const disabled = props.pathname ? props.pathname.includes(props.href) : false

  return (
    <button
      title="Expand Section"
      disabled={disabled}
      {...props}
      onMouseDown={(e) => e.preventDefault()}
      sx={{
        appearance: 'none',
        display: 'flex',
        alignItems: 'center',
        p: 2,
        m: 0,
        border: 0,
        borderRadius: 0,
        color: 'inherit',
        bg: 'transparent',
        ':hover,:focus': {
          color: 'primary',
        },
        '&:disabled': {
          opacity: 0.25,
        },
      }}
    >
      <svg viewBox="0 0 16 16" width="12" height="12">
        <g
          sx={{
            transformOrigin: '8 8',
            transition: 'transform .1s ease-out',
          }}
          transform={transform}
        >
          <path
            stroke="currentcolor"
            strokeWidth="2"
            fill="none"
            d="M14 6 L8 12 L2 6"
          />
        </g>
      </svg>
    </button>
  )
}

const NavLinks: FunctionComponent<{
  open: boolean
  pathname: string
  links: ReactComponentElement<'a'>[]
  href: string
  Link: React.FunctionComponent<{
    href?: string
    className?: string
    children?: ReactNode
  }>
}> = ({ open, pathname = '', links, href, Link }) => {
  if (!links) return null
  if (!open && !pathname.includes(href)) return null
  return (
    <ul
      sx={{
        listStyle: 'none',
        m: 0,
        p: 0,
      }}
    >
      {links.map((link, j) => (
        <li key={j}>
          <Link
            href={link.props.href}
            children={link.props.children}
            className={link.props.className}
            sx={{
              pl: 4,
            }}
          />
        </li>
      ))}
    </ul>
  )
}

export const AccordionNav = forwardRef<
  HTMLDivElement,
  {
    open?: boolean
    components?: MDXProviderComponents
    className?: string
    pathname?: string
    children?: ReactNode
  }
>(
  (
    { open, children, components = {}, className, pathname = '', ...props },
    ref
  ) => {
    const links = createNestedLinks(children)
    const [expanded, setExpanded] = useState<{ [k: number]: boolean }>({})
    const Link = components.a || 'a'

    const toggle = (i: number) => (e: MouseEvent) => {
      e.stopPropagation()
      setExpanded({
        ...expanded,
        [i]: !expanded[i],
      })
    }

    return (
      <React.Fragment>
        {open && <Overlay {...props} />}
        <div
          ref={ref}
          className={className}
          sx={{
            position: ['fixed', 'sticky'],
            top: 0,
            left: 0,
            bottom: [0, 'auto'],
            zIndex: 1,
            minWidth: 0,
            width: 256,
            maxHeight: ['100vh', 'none'],
            overflowX: 'visible',
            overflowY: 'auto',
            transition: 'transform .2s ease-out',
            transform: [open ? 'translateX(0)' : 'translate(-100%)', 'none'],
            bg: ['background', 'transparent'],
            WebkitOverflowScrolling: 'touch',
          }}
        >
          <ul
            sx={{
              listStyle: 'none',
              p: 0,
              m: 0,
            }}
          >
            {links.map((link: any, i: number) => (
              <li key={i}>
                <div
                  sx={{
                    display: 'flex',
                    alignItems: 'center',
                  }}
                >
                  <Link
                    href={link.props.href}
                    children={link.props.children}
                    className={link.props.className}
                  />
                  {link.props.links && (
                    <AccordionButton
                      href={link.props.href}
                      pathname={pathname}
                      open={expanded[i]}
                      sx={{
                        ml: 'auto',
                      }}
                      onClick={toggle(i)}
                    />
                  )}
                </div>
                <NavLinks
                  {...link.props}
                  open={expanded[i]}
                  pathname={pathname}
                  Link={Link}
                />
              </li>
            ))}
          </ul>
        </div>
      </React.Fragment>
    )
  }
)

const removeSlash = (str: string) =>
  str.length > 1 ? str.replace(/\/$/, '') : str

interface PaginationLinkProps {
  label: string
  mdxType: string
  originalType: string
  parentName: string
  children: ReactNode
}

const PaginationLink = ({
  label,
  children,
  mdxType,
  originalType,
  parentName,
  ...props
}: PaginationLinkProps) => (
  <a
    {...props}
    sx={{
      fontSize: 1,
      color: 'inherit',
      textDecoration: 'none',
      fontWeight: 'bold',
    }}
  >
    <div>{label}</div>
    <div
      sx={{
        fontSize: 3,
      }}
    >
      {children}
    </div>
  </a>
)

export interface PaginationProps {
  pathname: string
  components?: any
  children?: ReactNode
}

export const Pagination = ({
  pathname = '',
  children,
  components,
  ...props
}: PaginationProps) => {
  const links = flattenLinks(children)
  const index = links.findIndex(
    (link) =>
      React.isValidElement(link) && link.props.href === removeSlash(pathname)
  )
  const hasPagination = index > -1
  const previous = links[index - 1]
  const next = links[index + 1]

  return (
    <div
      {...props}
      sx={{
        display: 'flex',
      }}
    >
      {hasPagination && previous && React.isValidElement(previous) && (
        <PaginationLink {...previous.props} label="Previous:" />
      )}
      <div sx={{ mx: 'auto' }} />
      {hasPagination && next && React.isValidElement(next) && (
        <PaginationLink {...next.props} label="Next:" />
      )}
    </div>
  )
}