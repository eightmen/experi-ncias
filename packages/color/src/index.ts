import * as P from 'polished'
import { ColorModesScale, get, Theme } from '@theme-ui/css'

type Color = ColorModesScale[string]

/**
 * Get color from theme.colors
 */
export const getColor = (theme: Theme, color: Color) => {
  if (typeof color === 'object') {
    color = Array.isArray(color)
      ? color[0]
      : (color as Exclude<typeof color, string & {}>).__default
  }

  if (process.env.NODE_ENV !== 'production') {
    if (color && /^var\(--theme-ui-colors-(.+)\)$/.test(color)) {
      throw new Error(
        'A CSS property was passed to `getColor`. ' +
          '`theme.colors` contains CSS custom properties. ' +
          'Use `theme.rawColors` instead.'
      )
    }
  }

  return get(
    theme,
    'rawColors' in theme ? `rawColors.${color}` : `colors.${color}`,
    color
  )
}

/**
 * Darken a color by an amount 0–1
 */
export const darken = (c: Color, n: number) => (t: Theme) =>
  P.darken(n, getColor(t, c))
/**
 * Lighten a color by an amount 0–1
 */
export const lighten = (c: Color, n: number) => (t: Theme) =>
  P.lighten(n, getColor(t, c))
/**
 * Rotate the hue of a color by an amount 0–360
 */
export const rotate = (c: Color, d: number) => (t: Theme) =>
  P.adjustHue(d, getColor(t, c))

/**
 * Set the hue of a color to a degree 0–360
 */
export const hue = (c: Color, h: number) => (t: Theme) =>
  P.setHue(h, getColor(t, c))
/**
 * Set the saturation of a color to an amount 0–1
 */
export const saturation = (c: Color, s: number) => (t: Theme) =>
  P.setSaturation(s, getColor(t, c))
/**
 * Set the lightness of a color to an amount 0–1
 */
export const lightness = (c: Color, l: number) => (t: Theme) =>
  P.setLightness(l, getColor(t, c))
/**
 * Desaturate a color by an amount 0–1
 */
export const desaturate = (c: Color, n: number) => (t: Theme) =>
  P.desaturate(n, getColor(t, c))
/**
 * Saturate a color by an amount 0–1
 */
export const saturate = (c: Color, n: number) => (t: Theme) =>
  P.saturate(n, getColor(t, c))

/**
 * Shade a color by an amount 0–1
 */
export const shade = (c: Color, n: number) => (t: Theme) =>
  P.shade(n, getColor(t, c))
/**
 * Tint a color by an amount 0–1
 */
export const tint = (c: Color, n: number) => (t: Theme) =>
  P.tint(n, getColor(t, c))

export const transparentize = (c: Color, n: number) => (t: Theme) =>
  P.transparentize(n, getColor(t, c))
/**
 * Set the transparency of a color to an amount 0-1
 */
export const alpha = (c: Color, n: number) => (t: Theme) =>
  P.rgba(getColor(t, c), n)

/**
 * Mix two colors by a specific ratio
 */
export const mix =
  (a: Color, b: Color, n = 0.5) =>
  (t: Theme) =>
    P.mix(n, getColor(t, a), getColor(t, b))

/**
 * Get the complement of a color
 */
export const complement = (c: Color) => (t: Theme) =>
  P.complement(getColor(t, c))

/**
 * Get the inverted color
 */
export const invert = (c: Color) => (t: Theme) => P.invert(getColor(t, c))

/**
 * Desaturate the color to grayscale
 */
export const grayscale = (c: Color) => desaturate(c, 1)
