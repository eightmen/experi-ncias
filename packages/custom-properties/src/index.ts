import pluralize from 'pluralize'
import { Theme } from '@theme-ui/css'

// Simplified validator based on spec https://www.w3.org/TR/CSS22/syndata.html#value-def-identifier
// Does not check for "cannot start with a digit, two hyphens, or a hyphen followed by a digit"
const keyValidator = /^[A-z0-9][\w-]*$/

interface CustomProperties {
  [key: string]: string | number
}

export default function makeCustomProperties(theme: Theme, prefix?: string) {
  const customProperties: CustomProperties = {}

  const generateProperties = (object: object, previousKey?: string) => {
    Object.entries(object).forEach(([key, value]) => {
      let formattedKey = pluralize(key, 1) || key

      if (
        process.env.NODE_ENV !== 'production' &&
        !keyValidator.test(formattedKey)
      ) {
        console.warn(
          `[theme-ui] Theme key "${value}" found will produce an invalid CSS custom property. ` +
            'Keys must only contain the following: A-Z, a-z, 0-9, hyphen, underscore.'
        )
      }

      if (prefix && !previousKey) {
        formattedKey = `${prefix}-${formattedKey}`