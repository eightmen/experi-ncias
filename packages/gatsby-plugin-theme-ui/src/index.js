import { makeTheme } from '@theme-ui/css/utils'

// add colors, fonts, fontSizes, space, etc
const theme = makeTheme({})

export default theme
