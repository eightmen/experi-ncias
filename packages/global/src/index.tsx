import { jsx, type ThemeUIStyleObject } from '@theme-ui/core'
import { css, type Theme } from '@theme-ui/css'
import { Global as EmotionGlobal } from '@emotion/react'

export interface GlobalProps {
  styles: ThemeUIStyleObject
}
const Global = ({ styles }: GlobalProps): JSX.Element =>
  jsx(EmotionGlobal, {
    styles: (emotionTheme: unknown) => {
      const theme = emotionTheme as Theme
      return css(styles)(theme)
    },
  })

export default Global
