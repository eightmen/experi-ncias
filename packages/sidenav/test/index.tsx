
/**
 * @jest-environment jsdom
 * @jsx jsx
 */

import { jsx } from 'theme-ui'
import renderer from 'react-test-renderer'
import { render, cleanup, act } from '@theme-ui/test-utils'
import { ReactNode } from 'react'

import { Sidenav, Pagination, AccordionNav } from '../src'

afterEach(cleanup)

const Link = (props: { href: string; mdxType: 'a'; children: ReactNode }) => {
  const { children, ...rest } = props
  return <a {...rest}>{children}</a>
}

const Ul = (props: { mdxType: 'ul'; children: ReactNode }) => {
  const { children, ...rest } = props
  return <ul {...rest}>{children}</ul>
}

const links = (
  <ul>
    <li>
      <Link mdxType="a" href="/">
        Beep
      </Link>
    </li>
    <li>
      <Link mdxType="a" href="/boop">
        Boop
      </Link>
    </li>
    <li>
      <Link mdxType="a" href="/bop">
        Bop
      </Link>
    </li>
  </ul>
)
const nestedLinks = (
  <ul>
    <li>
      <Link mdxType="a" href="/">
        Beep
      </Link>
    </li>
    <li>
      <Link mdxType="a" href="/boop">
        Boop
      </Link>
    </li>
    <li>
      <Link mdxType="a" href="/bop">
        Bop
      </Link>
      <Ul mdxType="ul">
        <li>
          <Link mdxType="a" href="/bop/hi">
            Hi
          </Link>
        </li>
      </Ul>
    </li>
  </ul>
)

test('Sidenav renders', () => {
  const json = renderer.create(<Sidenav>{links}</Sidenav>).toJSON()
  expect(json).toMatchSnapshot()
})

test('Sidenav renders open', () => {
  const json = renderer.create(<Sidenav open={true}>{links}</Sidenav>).toJSON()
  expect(json).toMatchSnapshot()
})

test('Pagination renders', () => {
  const json = renderer
    .create(<Pagination pathname="/" children={links} />)
    .toJSON()
  expect(json).toMatchSnapshot()
})

test('AccordionNav renders', () => {
  const json = renderer.create(<AccordionNav children={links} />).toJSON()
  expect(json).toMatchSnapshot()
})

test('AccordionNav renders with nested links', () => {
  const json = renderer.create(<AccordionNav children={nestedLinks} />).toJSON()
  expect(json).toMatchSnapshot()
})

test('AccordionNav renders open', () => {
  const json = renderer
    .create(<AccordionNav open={true} children={nestedLinks} />)
    .toJSON()
  expect(json).toMatchSnapshot()
})

test('AccordionNav button toggles open state', () => {
  const root = render(<AccordionNav children={nestedLinks} />)
  const button = root.getByTitle('Expand Section')
  act(() => {
    button.click()
  })
  const nested = root.getByText('Hi')
  expect(nested).toBeTruthy()
})