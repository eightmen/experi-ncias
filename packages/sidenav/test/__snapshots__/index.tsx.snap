// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionNav renders 1`] = `
.emotion-0 {
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  z-index: 1;
  min-width: 0;
  width: 256px;
  max-height: 100vh;
  overflow-x: visible;
  overflow-y: auto;
  -webkit-transition: -webkit-transform .2s ease-out;
  transition: transform .2s ease-out;
  -webkit-transform: translate(-100%);
  -moz-transform: translate(-100%);
  -ms-transform: translate(-100%);
  transform: translate(-100%);
  background-color: background;
  -webkit-overflow-scrolling: touch;
}

@media screen and (min-width: 40em) {
  .emotion-0 {
    position: -webkit-sticky;
    positi